File: examples/scripts/ts/archive_session.ts
/**
 * Model Health TypeScript SDK — session archive.
 * Mirrors examples/python/archive_session.py.
 *
 * Usage:
 *   npx tsx archive_session.ts [<api_key>]
 */

import { ModelHealthClient } from '@modelhealth/modelhealth';
import type { Archive } from '@modelhealth/modelhealth';
import { loadApiKey, pickOne, confirm, saveFile, sleep, closePrompts } from './_shared.js';

async function connect(apiKey: string): Promise<ModelHealthClient> {
  console.log('Connecting...');
  const client = new ModelHealthClient({ apiKey, autoInit: false });
  await client.init();
  return client;
}

async function pickSession(client: ModelHealthClient) {
  console.log('\nFetching sessions...');
  const sessions = await client.sessionList();
  if (!sessions.length) {
    console.error('No sessions found. Create a session using the Model Health mobile app first.');
    process.exit(1);
  }

  console.log(`\n${sessions.length} session(s):\n`);
  return pickOne(sessions, 'Select session to archive', s => {
    const sn = s.sessionName || '(unnamed)';
    const sub = s.name || '(unnamed)';
    const actWord = s.activitiesCount === 1 ? 'activity' : 'activities';
    return `[session ID: ${s.id}]  session name: ${sn}  subject: ${sub}  ${s.activitiesCount} ${actWord}`;
  });
}

async function prepareArchive(
  client: ModelHealthClient,
  session: Awaited<ReturnType<typeof pickSession>>,
  withVideos: boolean
): Promise<Archive> {
  console.log(`\nRequesting archive for session '${session.id}'...`);
  const archive = await client.prepareArchive(session, withVideos);

  console.log('Waiting for archive to be ready...');
  const status = await pollArchive(client, archive);

  if (status.type !== 'ready') {
    console.error(`Archive preparation did not complete (status: ${status.type}).`);
    process.exit(1);
  }
  console.log('Archive is ready.');

  return archive;
}

async function downloadArchive(
  client: ModelHealthClient,
  archive: Archive,
  session: Awaited<ReturnType<typeof pickSession>>
) {
  console.log('\nDownloading...');
  const data = await client.archiveData(archive);
  const p = saveFile(`ModelHealth_Session_${session.id}.zip`, data);
  console.log(`  Saved ${p}  (${data.length.toLocaleString()} bytes)`);
}

async function main() {
  const args = process.argv.slice(2);
  const client = await connect(loadApiKey(args[0]));
  const session = await pickSession(client);

  console.log();
  const withVideos = await confirm('Include raw video files in the archive?', false);

  const archive = await prepareArchive(client, session, withVideos);
  await downloadArchive(client, archive, session);
  console.log('\nDone.');
}

async function pollArchive(client: ModelHealthClient, archive: Archive) {
  while (true) {
    const status = await client.archiveStatus(archive);
    if (status.type === 'processing') {
      process.stdout.write('  Preparing archive...  \r');
    } else {
      process.stdout.write('\n');
      return status;
    }
    await sleep(2_000);
  }
}

main().catch(err => { console.error(`Error: ${err.message ?? err}`); process.exit(1); })
  .finally(closePrompts);

File: examples/scripts/python/archive_session.py
#!/usr/bin/env python3
"""Model Health Python SDK — session archive demo.

Walks through the archive workflow:
  1. Select a session
  2. Choose whether to include raw video files
  3. Request archive preparation
  4. Poll until the archive is ready (or fails)
  5. Download and save the ZIP file

Usage:
    archive_session.py [<api_key>]
"""

import sys
import time

from docopt import docopt

from modelhealth import (
    ModelHealthService,
    ModelHealthError,
    ArchiveStatus,
)
from _prompts import pick_one, confirm
from _utils import save_file, load_api_key

# ---------------------------------------------------------------------------
# Polling helper
# ---------------------------------------------------------------------------

def _poll_archive(service, archive, interval=2):
    """Block until the archive is ready or has failed.

    Returns the final status (ArchiveStatus.ready or ArchiveStatus.failed).
    """
    while True:
        status = service.archive_status(archive)
        if status == ArchiveStatus.processing:
            print("  Preparing archive...  ", end="\r", flush=True)
        else:
            print()  # clear the \r line
            return status
        time.sleep(interval)


# ---------------------------------------------------------------------------
# Setup
# ---------------------------------------------------------------------------

def _connect(api_key):
    print("Connecting...")
    try:
        return ModelHealthService(api_key)
    except ModelHealthError as exc:
        sys.exit(f"Failed to initialise: {exc}")


# ---------------------------------------------------------------------------
# Session selection
# ---------------------------------------------------------------------------

def _pick_session(service):
    print("\nFetching sessions...")
    sessions = service.session_list()
    if not sessions:
        sys.exit(
            "No sessions found. Create a session using the Model Health mobile app first."
        )

    print(f"\n{len(sessions)} session(s):\n")
    return pick_one(
        sessions,
        "Select session to archive",
        lambda s: f"[session ID: {s.id}]  session name: {s.session_name or '(unnamed)'}  subject: {s.name or '(unnamed)'}  {s.activities_count} {'activity' if s.activities_count == 1 else 'activities'}",
    )


# ---------------------------------------------------------------------------
# Archive preparation
# ---------------------------------------------------------------------------

def _prepare_archive(service, session, with_videos):
    print(f"\nRequesting archive for session '{session.id}'...")
    try:
        archive = service.prepare_archive(session, with_videos=with_videos)
    except ModelHealthError as exc:
        sys.exit(f"Failed to start archive preparation: {exc}")

    print("Waiting for archive to be ready...")
    status = _poll_archive(service, archive)

    if status != ArchiveStatus.ready:
        sys.exit(f"Archive preparation did not complete (status: {status}).")
    print("Archive is ready.")

    return archive


# ---------------------------------------------------------------------------
# Download
# ---------------------------------------------------------------------------

def _download_archive(service, archive, session):
    print("\nDownloading...")
    try:
        data = service.archive_data(archive)
    except ModelHealthError as exc:
        sys.exit(f"Failed to download archive: {exc}")

    path = save_file(f"ModelHealth_Session_{session.id}.zip", data)
    print(f"  Saved {path}  ({len(data):,} bytes)")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main(api_key):
    service = _connect(api_key)
    session = _pick_session(service)

    print()
    with_videos = confirm("Include raw video files in the archive?", default=False)

    archive = _prepare_archive(service, session, with_videos)
    _download_archive(service, archive, session)
    print("\nDone.")


if __name__ == "__main__":
    args = docopt(__doc__)
    main(load_api_key(args["<api_key>"]))
